mcpserver/tools.go

```diff
@@ -461,68 +484,81 @@ Failure to do so will result in the tool being stuck, awaiting for the command t
-				}
+		return req, nil
+	},
+	ActualHandler: func(ctx context.Context, request *EnvironmentRunCmdToolRequest) (*ToolResponse[environment.EndpointMappings], error) {
+		repo, env, err := openEnvironmentFromRequest(ctx, request.BaseEnvironmentRequest)
+		if err != nil {
+			return nil, fmt.Errorf("unable to open the environment: %w", err)
+		}
+
+		updateRepo := func() error {
+			if err := repo.Update(ctx, env, request.Explanation); err != nil {
+				return fmt.Errorf("failed to update repository: %w", err)
 			}
-			endpoints, runErr := env.RunBackground(ctx, command, shell, ports, request.GetBool("use_entrypoint", false))
+			return nil
+		}
+
+		if request.Background {
+			endpoints, runErr := env.RunBackground(ctx, request.Command, request.Shell, request.Ports, request.UseEntrypoint)
 			// We want to update the repository even if the command failed.
-			if resp, err := updateRepo(); err != nil {
-				return resp, nil
+			if err := updateRepo(); err != nil {
+				return nil, err
 			}
 			if runErr != nil {
-				return mcp.NewToolResultErrorFromErr("failed to run command", runErr), nil
+				return nil, fmt.Errorf("failed to run command: %w", runErr)
 			}
 
-			out, err := json.Marshal(endpoints)
-			if err != nil {
-				return nil, err
-			}
-
-			return mcp.NewToolResultText(fmt.Sprintf(`Command started in the background in NEW container. Endpoints are %s
+			return &ToolResponse[environment.EndpointMappings]{
+				Data: endpoints,
+				Message: fmt.Sprintf(`Command started in the background in NEW container.
 
 To access from the user's machine: use host_external. To access from other commands in this environment: use environment_internal.
 
 Any changes to the container workdir (%s) WILL NOT be committed to container-use/%s
 
 Background commands are unaffected by filesystem and any other kind of changes. You need to start a new command for changes to take effect.`,
-				string(out), env.Config.Workdir, env.ID)), nil
+					env.Config.Workdir, env.ID),
+			}, nil
 		}
 
-		stdout, runErr := env.Run(ctx, command, shell, request.GetBool("use_entrypoint", false))
+		stdout, runErr := env.Run(ctx, request.Command, request.Shell, request.UseEntrypoint)
 		// We want to update the repository even if the command failed.
-		if resp, err := updateRepo(); err != nil {
-			return resp, nil
+		if err := updateRepo(); err != nil {
+			return nil, err
 		}
 		if runErr != nil {
-			return mcp.NewToolResultErrorFromErr("failed to run command", runErr), nil
+			return nil, fmt.Errorf("failed to run command: %w", runErr)
 		}
 
-		return mcp.NewToolResultText(fmt.Sprintf("%s\n\nAny changes to the container workdir (%s) have been committed and pushed to container-use/ remote", stdout, env.Config.Workdir)), nil
+		return &ToolResponse[environment.EndpointMappings]{
+			Message: fmt.Sprintf("%s\n\nAny changes to the container workdir (%s) have been committed and pushed to container-use/ remote", stdout, env.Config.Workdir),
+		}, nil
```

Comment:
🐛 **Bug (Medium):** Missing Data field in ToolResponse for non-background command execution. Should provide appropriate data or adjust return type.

```suggestion
		return &ToolResponse[environment.EndpointMappings]{
			Data:    environment.EndpointMappings{}, // or appropriate empty value
			Message: fmt.Sprintf("%s\n\nAny changes to the container workdir (%s) have been committed and pushed to container-use/ remote", stdout, env.Config.Workdir),
		}, nil
```